packages/firestore/src/util/misc.ts

```diff
@@ -74,6 +75,48 @@ export interface Equatable<T> {
   isEqual(other: T): boolean;
 }
 
+/** Compare strings in UTF-8 encoded byte order */
+export function compareUtf8Strings(left: string, right: string): number {
+  let i = 0;
+  while (i < left.length && i < right.length) {
+    const leftCodePoint = left.codePointAt(i)!;
+    const rightCodePoint = right.codePointAt(i)!;
+
+    if (leftCodePoint !== rightCodePoint) {
+      if (leftCodePoint < 128 && rightCodePoint < 128) {
+        // ASCII comparison
+        return primitiveComparator(leftCodePoint, rightCodePoint);
+      } else {
+        // Lazy instantiate TextEncoder
+        const encoder = newTextEncoder();
+
+        // UTF-8 encoded byte comparison, substring 2 indexes to cover surrogate pairs
+        const leftBytes = encoder.encode(left.substring(i, i + 2));
+        const rightBytes = encoder.encode(right.substring(i, i + 2));
+        for (
+          let j = 0;
+          j < Math.min(leftBytes.length, rightBytes.length);
+          j++
+        ) {
+          const comparison = primitiveComparator(leftBytes[j], rightBytes[j]);
+          if (comparison !== 0) {
+            return comparison;
+          }
+        }
+
+        // Compare lengths if all bytes are equal
```

Comment:
Under what circumstances would all bytes be equal? If there were all equal, then wouldn't the "if" condition on line 85 have evaluated to false?

Reply:
yeah, IIUC since `leftCodePoint !== rightCodePoint`, `comparison` on line 101 should be non-zero for at least one iteration of the loop on line 96.

Reply:
nope, there are some cases where trailing surrogate gets falsy involved, and being misrepresented by bytes, which could lead to 2 equal byte array for 2 different code points. That's why the unit tests were failing in android sdk.

Reply:
😮

Reply:
> which could lead to 2 equal byte array for 2 different code points

I'm having a very hard time understanding this.

A code point is just a number (at least in java it's [just an int](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#codePointAt-char:A-int-)). How can 2 identical integers result in 2 different series of bits?

Reply:
I am willing to show the edge case and brainstorm if u would like to discuss this offline.

Reply:
I'd recommend making this comparison logic robust in the presence of invalid UTF-16 strings. Otherwise, there is a potential for a DoS attack by providing invalid UTF-16 strings. That being said, robustness in handling invalid UTF-16 could be added in a follow-up PR. If you choose to defer to a later PR, please file a buganizer ticket and add a `// TODO(b/NNNNNN): handle invalid UTF-16 surrogates` comment.

A different, but similar, issue was exploited in "overlong encodings" in utf8 to bypass security checks: https://en.wikipedia.org/wiki/UTF-8#Overlong_encodings

Reply:
The unit tests includes invalid UTF-16 surrogates, and with the current code, both integration and unit tests are passing as expected. Can you elaborate on the "robustness in handling invalid UTF-16"?

Reply:
> like throwing an error message instead of decoding it?

Yes, that is what I'm suggesting. I imagine doing so will make the code simpler and easier to understand. The logic in this PR is incredibly complex and non-intuitive, so anything to make it more readable is a "win" IMO. Using a follow-up PR to add improvements like this shows the incremental development and enables digesting the logic in steps instead of all at once. This is, however, merely a _suggestion_. Choice is up to you.